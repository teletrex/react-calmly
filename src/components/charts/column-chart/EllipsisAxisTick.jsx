/* <LICENSE>
* 
* Copyright (C) 2022 Louis F. Roehrs, All rights reserved.
* 
* </LICENSE>
*  */



import React, { useCallback, useMemo, useState } from 'react';
import PropTypes from 'prop-types';
import { Text } from 'recharts';
import maxBy from 'lodash/maxBy';

import { axisTypes } from './utilities';

/*
https://github.com/recharts/recharts/issues/961
This component is responsible for truncating ticks inside the rechart 
charts in both quantity of lines and maximum width for single line.

maxLines - used for limiting amount of lines in single tick
width - used for limiting maximum length for single world
*/

const EllipsisAxisTick = ({ axis, labelAngle, maxLines, payload, tickFormatter, ...rest }) => {
  const [text, setText] = useState(payload.value);
  const [suffix, setSuffix] = useState('');

  const formattedValue = useMemo(
    () => (tickFormatter ? tickFormatter(payload.value) : payload.value),
    [payload.value]
  );

  const measuredRef = useCallback(
    node => {
      if (node === null) {
        return;
      }

      const getWordsByLines = (node, children) => node.getWordsByLines({ ...rest, children }, true);

      let tempText = formattedValue?.toString() ?? '';

      let wordsByLines = getWordsByLines(node, tempText);
      let longestLine = maxBy(wordsByLines, 'width');

      const calculatedAngle = labelAngle * (Math.PI / 180);
      const wordInclination =
        axis === axisTypes.x
          ? Math.abs(Math.sin(calculatedAngle))
          : Math.abs(Math.cos(calculatedAngle));
      const axisLabelSpace = axis === axisTypes.x ? rest.height : rest.width;
      const wordLengthLimit = axisLabelSpace / wordInclination;
      const tempSuffix =
        wordsByLines.length > maxLines || longestLine.width > wordLengthLimit ? 'â€¦' : '';

      while (wordsByLines.length > maxLines || (longestLine.width > wordLengthLimit && tempText)) {
        tempText = tempText.slice(0, -1);
        wordsByLines = getWordsByLines(node, tempText + tempSuffix);
        longestLine = maxBy(wordsByLines, 'width');
      }

      if (tempText !== text) {
        setText(tempText);
        setSuffix(tempSuffix);
      }
    },
    [axis, labelAngle, maxLines, rest, text, payload]
  );

  return (
    <g>
      <Text {...rest} ref={measuredRef}>
        {text + suffix}
      </Text>
      <title>{payload.value}</title>
    </g>
  );
};

EllipsisAxisTick.defaultProps = {
  ...Text.defaultProps,
  axis: axisTypes.x,
  payload: {},
  tickFormatter: undefined,
};

EllipsisAxisTick.propTypes = {
  ...Text.propTypes,
  axis: PropTypes.oneOf(Object.values(axisTypes)).isRequired,
  labelAngle: PropTypes.number,
  maxLines: PropTypes.number.isRequired,
  payload: PropTypes.shape({
    value: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  }),
  tickFormatter: PropTypes.func,
};

export default EllipsisAxisTick;
