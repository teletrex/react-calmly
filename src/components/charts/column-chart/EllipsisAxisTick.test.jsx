/* <LICENSE>
* 
* Copyright (C) 2022 Louis F. Roehrs, All rights reserved.
* 
* </LICENSE>
*  */



import React from 'react';
import { render } from '@testing-library/react';

import { mockGetBoundingClientRect } from '../../utils/test-helpers/html-element-mocks';

import EllipsisAxisTick from './EllipsisAxisTick';

describe('EllipsisAxisTick', () => {
  beforeEach(() => {
    mockGetBoundingClientRect({ width: 10 });
  });
  afterEach(() => {
    jest.restoreAllMocks();
  });

  const getDefaultProps = overrides => ({
    fontSize: 11,
    lineHeight: '1em',
    textAnchor: 'end',
    scaleToFit: false,
    width: 50,
    maxLines: 1,
    ...overrides,
  });

  const labels = [['2019 Q1'], [undefined], [null], [150000000]];

  test.each(labels)('verify if created when the value is %p', value => {
    const { container } = render(
      <svg>
        <EllipsisAxisTick {...getDefaultProps({ payload: { value } })} />
      </svg>
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  test('verify if label is given ellipsis if it is longer then width', () => {
    const { container } = render(
      <svg>
        <EllipsisAxisTick
          {...getDefaultProps({
            payload: { value: 'Lorem ipsum dolor sit amet, consectetur adipiscing elit' },
          })}
        />
      </svg>
    );

    const axisLabelValue = container.querySelector('tspan').textContent;

    expect(axisLabelValue).toBe('Lorem ipsumâ€¦');
  });

  test('verify if label is formatted by tickFormatter', () => {
    const { container } = render(
      <svg>
        <EllipsisAxisTick
          {...getDefaultProps({
            payload: { value: 10000 },
            tickFormatter: value => `${value}$`,
          })}
        />
      </svg>
    );

    const axisLabelValue = container.querySelector('tspan').textContent;

    expect(axisLabelValue).toBe('10000$');
  });
});
