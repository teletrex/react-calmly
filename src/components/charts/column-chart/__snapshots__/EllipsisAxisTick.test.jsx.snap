// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EllipsisAxisTick verify if created when the value is "2019 Q1" 1`] = `
<svg>
  <g>
    <text
      class="recharts-text"
      font-size="11"
      text-anchor="end"
      width="50"
      x="0"
      y="0"
    >
      <tspan
        dy="0em"
        x="0"
      >
        2019 Q1
      </tspan>
    </text>
    <title>
      2019 Q1
    </title>
  </g>
</svg>
`;

exports[`EllipsisAxisTick verify if created when the value is 150000000 1`] = `
<svg>
  <g>
    <text
      class="recharts-text"
      font-size="11"
      text-anchor="end"
      width="50"
      x="0"
      y="0"
    >
      <tspan
        dy="0em"
        x="0"
      >
        150000000
      </tspan>
    </text>
    <title>
      150000000
    </title>
  </g>
</svg>
`;

exports[`EllipsisAxisTick verify if created when the value is null 1`] = `
<svg>
  <g>
    <text
      class="recharts-text"
      font-size="11"
      text-anchor="end"
      width="50"
      x="0"
      y="0"
    >
      <tspan
        dy="0em"
        x="0"
      />
    </text>
    <title />
  </g>
</svg>
`;

exports[`EllipsisAxisTick verify if created when the value is undefined 1`] = `
<svg>
  <g>
    <text
      class="recharts-text"
      font-size="11"
      text-anchor="end"
      width="50"
      x="0"
      y="0"
    >
      <tspan
        dy="0em"
        x="0"
      />
    </text>
    <title />
  </g>
</svg>
`;
